src/rulesets/Base/fastmath_able.jl

```diff
@@ -36,10 +36,88 @@ let
 
 
         # Unary complex functions
-        @scalar_rule abs(x::Real) sign(x)
-        @scalar_rule abs2(x) 2x
-        @scalar_rule angle(x::Real) Zero()
-        @scalar_rule conj(x::Real) One()
+        ## abs
+        function frule((_, Δx), ::typeof(abs), x::Real)
+            return abs(x), sign(x) * real(Δx)
+        end
+        function frule((_, Δz), ::typeof(abs), z::Complex)
+            Ω = abs(z)
+            return Ω, (real(z) * real(Δz) + imag(z) * imag(Δz)) / Ω
+        end
+        
+        function rrule(::typeof(abs), x::Real)
+            function abs_pullback(Δf)
+                return (NO_FIELDS, real(Δf)*sign(x))
+            end
+            return abs(x), abs_pullback
+        end
+        function rrule(::typeof(abs), z::Complex)
+            Ω = abs(z)
+            function abs_pullback(Δf)
+                Δu = real(Δf) 
+                return (NO_FIELDS, Δu*z/Ω)
```

Comment:
Should we special-case when `z=0` here? e.g. for `sign`, the pushforward/pullback is technically undefined at `z=0`, but following the subgradient convention, they are given a value of `0`.

```suggestion
                return (NO_FIELDS, Δu*z/ifelse(iszero(z), one(Ω), Ω)
```

Reply:
Hm. I'm not very familiar with this 'subgradient convention', but I must say it makes me uneasy. Personally, I think it'd rather just encounter a `NaN` than get zero gradient, but I can see how producing `NaN` willy-nilly in an AD system could be frustrating.

Reply:
@oxinabox wrote a little bit about it here: https://github.com/JuliaDiff/ChainRules.jl/pull/185#discussion_r421675694.

Reply:
The short version of the subgradient convention is that we claim derivative is zero at local minima, regardless of if it is otherwise differentiable or not.
Because its useful to people doing optimization.
Its not something i feel that strongly about though.

For real numbers the common alternative in AD is to instead chose 1 or -1, which would be the program deriviative if going through control flow,
and the function had being implmented wioth a `if x <= 0` etc.
Which I think while acceptable is just worse.

For complex numbers, I don't know.
I am not sure the optimization usecase even applies.

I will not block this PR over this.

Reply:
Yeah, I found a few different conventions in ChainRules (e.g. `mod` uses `NaN` where it is not differentiable). So I also will not block over this.

Reply:
Optimization examples certainly do apply, especially in C -> R functins like `abs`. After some reflection, I think I can get behind using this subgradient convention here. I'll make the appropriate changes.

Reply:
@sethaxen Is there a reason you recommend `Δu*z/ifelse(iszero(z), one(Ω), Ω)` rather than 
```juia
ifelse(iszero(z), zero(Ω), Δu*z/Ω)
```
? I guess your version is better behaved under promotion?

Reply:
> I guess your version is better behaved under promotion?

Yeah that's right, it makes it type-stable.